content/en/docs/performance-workflow.md

```diff
@@ -6,11 +6,103 @@ description: >-
+  
+You will see that the queries have a the traceback which will lead you to the exact section of code causing the issue.  
+Here the issue is due to the code in `app/controllers/listworkers_controller.rb` :
+```
+@p_cat_choosen = Array.new
+@tab_cat.each do |category_id|
+  unless category_id.blank?
+    s_category = SCategory.find(category_id)
+    @p_cat_choosen << s_category.p_category unless @p_cat_choosen.include?(s_category.p_category)
+  end
+end
+```
+
+### Extract and Test
+
+Once you have identified the faulty code you should extract it into its own method.  
+In the previous example the extracted code was placed in `app/models/p_category.rb` :  
+```
+def self.get_p_categories_from_s_category_ids_deprecated(s_category_ids)
+  chosen_p_categories = Array.new
+  s_category_ids.each do |category_id|
+    unless category_id.blank?
+      s_category = SCategory.find(category_id)
+      chosen_p_categories << s_category.p_category unless chosen_p_categories.include?(s_category.p_category)
+    end
+  end
+  chosen_p_categories
+end
+```
+You should then thouroughly test the method in such a way as to ensure the modification will behave in the same way.
+
+
+### Improve Performance
+
+At this point you can now change the method name to include a _"deprecated"_ suffix.  
+You then re-write the method in a more performant way. Make sure the new code still passes on all the tests written previously.  
+
+```
+def self.get_p_categories_from_s_category_ids_deprecated(s_category_ids)
+  chosen_p_categories = Array.new
+  s_category_ids.each do |category_id|
+    unless category_id.blank?
+      s_category = SCategory.find(category_id)
+      chosen_p_categories << s_category.p_category unless chosen_p_categories.include?(s_category.p_category)
+    end
+  end
+  chosen_p_categories
+end
+
+def self.get_p_categories_from_s_category_ids(s_category_ids)
+  PCategory.joins(:s_categories).where('s_categories.id IN (?)', s_category_ids).distinct.to_a
+end
+```
+
+You now should write a comparative performance test. This can be placed in the `spec/performance` folder.  
+This should follow the following format:
+```
+it "should run quicker than the old implementation" do
+  expect { subject.tested_method }.to(perform_faster_than { subject.tested_method_deprecated }.at_least(2).times)
+end
+```
+Once the new method has passed all the tests and performs to the desired standard, you should commit you work.
+
+### Archive
+
+Finally it is time to archive your work. The archiving of the improvement is done in `spec/performance/PERFORMANCE_ARCHIVE.md`. For this the following information should be included:
+
+* Date of commit
+* Commit id for the change
+* How many times faster the new method is to the old one
+* The file location
+* The relevant code change
+  
+<img width="1034" alt="Screenshot 2021-04-16 at 15 47 07" src="https://user-images.githubusercontent.com/33180547/115033723-0be46800-9ecb-11eb-86f0-c11b55f87ad7.png">
```

Comment:
As discussed, I'd prefer to avoid this, and instead add a section to the best practices if there's a common knowledge to be learned from here in the form of: Prefer X over Y, Avoid using Z, Always try M before N ... or so